### ISSUE TITLE
Online learning, adding data to the network

### ISSUE THREAD

**Issue by oritwoen** [cazala/synaptic#131]
Hi, I'm sorry that it adds a question here but I just do not really get anywhere else answer.

After learning network and exporting it to JSON and then restored, you can add new data without re-learning the whole network?

I mean for example, where we learn our program underlying concepts. Then, after the introduction of the check function myNetwork.activate () when the answer will be 0 adding this data to the network to the next check myNetwork.activate () gave the answer 1.

**Comment by bobalazek** [comment]
Sure, why not? As long as you have the same configuration (same number of input, hidden and output layers).

**Comment by I-n-d-e-r** [comment]
Yes, it works for small values of input and output but normalize and teach in such a way that the text is difficult.

@bobalazek you can not use your algorithm example: https://github.com/bobalazek/synaptic-text-classification-example because every time the number of inputs and outputs are different so you can not upgrade the network only need to learn it from scratch.

I tried to use the following normalization algorithm and indeed work but it is difficult to speak of correct diagnoses because it is a binary value, and every word becomes too similar.

```js
var DataSet = [
{
      input: Normalize(Input),
      output: [0]
},
{
      input: Normalize(Input2),
      output: [1],
}]

this.Network = new Architect.Perceptron(100, 101, 1)
this.Trainer = new Trainer(this.Network)
this.Tests = this.Trainer.train(DataSet,
{
      rate: .1,
      iterations: 100000,
      error: .005,
      shuffle: true,
      log: 10,
      cost: Trainer.cost.CROSS_BINARY
})

Normalize = function(ascii)
{
      var bin = "0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000";
      for (var i = 0; i < ascii.length; i++)
      {
          var code = ascii.charCodeAt(i);
          bin += ('0000000000' + code.toString(2)).slice(-10);
      }
      return bin.slice(-10 * 10).split('').reverse();
}
```

**Comment by bobalazek** [comment]
Yes, that's correct. That's why I said that you CAN reuse it IF you have the same configuration (number of input, hidden and output layers). The input layers would be the number of words, and the output layer, the number of categories / classes -- with the bag-of-words model.

Oh yeah, I know, that is a pretty old example I have there. I have a much newer version currently local... will open-source it soon, when I perfected it -- far better than the old one.

I'm currently still fairly new with the NN stuff. But probably the best option is still to experiment stuff and see how it works out with different configuration and models. :)

**Comment by bobalazek** [comment]
Ok, I've made a "simple trainer" for synaptic, with the bag-of-words model. You can take a look here:
https://github.com/bobalazek/synaptic-simple-trainer

**Comment by cazala** [comment]
nice!

**Comment by I-n-d-e-r** [comment]
@bobazalek, thanks! It looks really good, I can use it in my project :)

**Comment by frollino** [comment]
please remove my mail from mailing list. thanks


Il 18/08/2016 19:52, Borut Balazek ha scritto:

**Comment by bobalazek** [comment]
@I-n-d-e-r great! For now, you can just look at the source code for the configuration (https://github.com/bobalazek/synaptic-simple-trainer/blob/master/lib/index.js). You just set any setting, BEFORE `trainer.prepareData()`, but I'll add some more documentation today.

@frollino I'm not really sure how you came to this list (not sure if there's another way as manually subscribing to it), but you can do that either by clicking the "or mute the thread link: ..." inside the email, or by clicking "unsubscribe" on the right side of this issue page (https://github.com/cazala/synaptic/issues/131).

**Comment by bobalazek** [comment]
Just added the functionality of continuous training (see example here: https://github.com/bobalazek/synaptic-simple-trainer/blob/master/example/train-new.js). Sadly it's not that optimal as we may want to, but I don't think that's there's another option (at least not with the bag-of-words model and a perceptron; well, at least I don't know if it is possible).
How it works:
* First you need to run the normal training
* Create a separate file for the new training (as in my example: `train-new.js`)
  * Set `trainer.continuousTraining = true;` and `trainer.trainData = data.train_new;` (it's important there, that you only set the new data here -- it will warn you if there are some duplicate texts, but you can easily override that with the `-f` flag at the end of the command. It's not that bad, but it's probably annoying to have duplicate entries in there)
* Run the new training -- if something is wrong, then you'll get an error

Important note: Like said, this will probably not be that optimal on smaller training sets, since there you can easily add new words -- one new word / class, and you'll need to retrain it from scratch again. On larger datasets you may have a better possibility, that it will work fine, since there, many words will already be used.

The only fix that I can think of in this case, would be, that you create a few blank inputs (words) and outputs (classes), that would act as placeholders for possible new words / classes, but that would only be a temporal solution, since you can't really make infinitive number of those, so you'll need to retrain the network sooner or later. Also, you really don't want to do that, since in large datasets, you'll see, that having many inputs and outputs it's a very time consuming training (I'm currenty doing a network with 800 inputs, 128 hidden and 200 output layers and it takes around 16 hours to train -- it would probably take days, if I would set the hidden layers, to a reasonable number, which in this case would be around 400). :)

**Comment by I-n-d-e-r** [comment]
As for me is great, of course, it is not so freely at certain times but just until it finds a better solution :) The only comment I have is that in addition to instead write to the file can be exported in their own project network if someone for example, wanted to use Redis or another database :)

**Comment by bobalazek** [comment]
Oh yeah, makes sense. Will add import / export adapters soon. If you have any other ideas, just write a issue on the repository and we can talk about it. :)
